File: app/backend/.net/SmartRide.AuthService/Controllers/ComplaintController.cs
using Microsoft.AspNetCore.Mvc;
using SmartRide.AuthService.Models.Entities;
using SmartRide.AuthService.Services;

namespace SmartRide.AuthService.Controllers;

[ApiController]
[Route("api/complaints")]
public class ComplaintController : ControllerBase
{
    private readonly IComplaintService _complaintService;

    public ComplaintController(IComplaintService complaintService)
    {
        _complaintService = complaintService;
    }

    [HttpGet]
    public async Task<ActionResult<List<Complaint>>> GetAllComplaints()
    {
        var complaints = await _complaintService.GetAllComplaintsAsync();
        return Ok(complaints);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetComplaintById([FromRoute] int id)
    {
        var complaint = await _complaintService.GetComplaintByIdAsync(id);
        if (complaint == null)
        {
            return NotFound();
        }
        return Ok(complaint);
    }

    [HttpGet("user/{userId}")]
    public async Task<ActionResult<List<Complaint>>> GetComplaintsByUserId([FromRoute] int userId)
    {
        var complaints = await _complaintService.GetComplaintsByUserIdAsync(userId);
        return Ok(complaints);
    }

    [HttpPost]
    public async Task<IActionResult> CreateComplaint([FromBody] Complaint complaint)
    {
        var savedComplaint = await _complaintService.CreateComplaintAsync(complaint);
        return CreatedAtAction(nameof(GetComplaintById), new { id = savedComplaint.ComplaintId }, savedComplaint);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateComplaintStatus([FromRoute] int id, [FromBody] Dictionary<string, string> payload)
    {
        try
        {
            payload.TryGetValue("status", out var status);
            payload.TryGetValue("resolutionNotes", out var resolutionNotes);

            var updatedComplaint = await _complaintService.UpdateComplaintStatusAsync(id, status, resolutionNotes);
            return Ok(updatedComplaint);
        }
        catch (Exception ex)
        {
            return NotFound(new { error = ex.Message });
        }
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteComplaint([FromRoute] int id)
    {
        try
        {
            await _complaintService.DeleteComplaintAsync(id);
            return Ok(new { message = "Complaint deleted successfully" });
        }
        catch (Exception ex)
        {
            return NotFound(new { error = ex.Message });
        }
    }
}

File: app/backend/java/auth-service/src/main/java/com/srbms/authservice/controller/ComplaintController.java
package com.srbms.authservice.controller;

import com.srbms.authservice.entity.Complaint;
import com.srbms.authservice.service.ComplaintService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/complaints")
public class ComplaintController {

    @Autowired
    private ComplaintService complaintService;

    @GetMapping({"", "/", "/all"})
    public ResponseEntity<List<Complaint>> getAllComplaints() {
        return ResponseEntity.ok(complaintService.getAllComplaints());
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getComplaintById(@PathVariable Integer id) {
        return complaintService.getComplaintById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<List<Complaint>> getComplaintsByUserId(@PathVariable Integer userId) {
        return ResponseEntity.ok(complaintService.getComplaintsByUserId(userId));
    }

    @PostMapping
    public ResponseEntity<Complaint> createComplaint(@RequestBody Complaint complaint) {
        Complaint savedComplaint = complaintService.createComplaint(complaint);
        return ResponseEntity.status(HttpStatus.CREATED).body(savedComplaint);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateComplaintStatus(
            @PathVariable Integer id,
            @RequestBody Map<String, String> payload) {
        try {
            String status = payload.get("status");
            String resolutionNotes = payload.get("resolutionNotes");
            Complaint updatedComplaint = complaintService.updateComplaintStatus(id, status, resolutionNotes);
            return ResponseEntity.ok(updatedComplaint);
        } catch (Exception e) {
            Map<String, String> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteComplaint(@PathVariable Integer id) {
        try {
            complaintService.deleteComplaint(id);
            Map<String, String> response = new HashMap<>();
            response.put("message", "Complaint deleted successfully");
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, String> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }
    }
}
